use only isito chart to deploy all components

luisyonaldo opened the pull request:
Use helm hooks to install CRD before installing the custom objects that depend on them. With this then there is not need to run `deployIstioHelmCharts.sh` to install istio, just `helm install istio` is enough.

TODO:

- [x]  Move all components into istio chart
- [x] Use post-install hook to install custom objects (that depend on CRD)
- [x] Move all needed manifests to `istio` chart
- [x] Full test

How to test it:
run `helm install istio` (on `install/kubernetes/helm` folder). 

helm hooks are not quite ready to handle CRD. Even if you define CRD as a `pre-install` hook, `helm isntall` will fail if it founds a custom object that depend on this CRD (so they use api "config.istio.io") in the same install. This is due to validation step that helm does before installing any object. Check kubernetes/helm#2925 (comment)

So, according to my knowledge, there two ways to install pass this issue using helm right now:

- A two install step:
First install CRD and then install the custom objects that depend on them. This is what the incubator's istio chart is doing with a `helm init` followed by a `helm upgrade` as you can see [here](https://github.com/kubernetes/charts/tree/master/incubator/istio#installing-the-chart).

- Install custom objects in a `post-install` hook:
In order to pass helm validation, you can create custom objects out of helm context by running a `post-intall` job that will manually create custom objects. This is what I choose in this PR, as you can see `create-custom-objects-job.yaml` job create custom object  by running `kubectl apply -f custom-objects` after helm installation. Also, this is what `prometheus-operator` is doing right now as you can see here in the [templates](https://github.com/coreos/prometheus-operator/tree/master/helm/prometheus-operator).

I'll open an issue in `helm` repo to see if there is a better way to do this. Let me know what you guys think and if you can install all by just running `helm install`.

istio-testing commented:
Hi @luisyonaldo. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should join the org to skip this step.

I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).

<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>

rshriram commented:
/ok-to-test

istio-testing commented:
@luisyonaldo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | b04b318d6feef74300deaaf83d5d6fdd5c0d5c6b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2461/istio-presubmit/2197/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

sdake commented:
As there is no check/gate on helm installation, please give me an opportunity to pull down this patch today and provide a lgtm if it works properly.

Cheers
-steve

istio-testing commented:
@luisyonaldo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 84fdaec5e79bafbb9abefc44db1e27a60d687aaf | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2461/istio-presubmit/2208/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

luisyonaldo commented:
@sdake the PR is ready to test, just run `helm install istio` on `install/kubernetes/helm` folder and it should install all istio components.

Since I added some [values](https://github.com/luisyonaldo/istio/blob/84fdaec5e79bafbb9abefc44db1e27a60d687aaf/install/kubernetes/templates/helm/istio/values.yaml.tmpl) to the chart, I modified [this](https://github.com/luisyonaldo/istio/blob/84fdaec5e79bafbb9abefc44db1e27a60d687aaf/pilot/cmd/istioctl/gendeployment/helm.go) and [this](https://github.com/luisyonaldo/istio/blob/84fdaec5e79bafbb9abefc44db1e27a60d687aaf/pilot/cmd/istioctl/gendeployment/helm.go) files, not sure if that break test.

ZackButcher commented on pilot/cmd/istioctl/gendeployment/helm.go:
```diff
@@ -40,6 +40,8 @@ global:
    proxy_debug: {{ .Debug }}
    pilot_hub: {{ .Hub }}
    pilot_tag: {{ .PilotTag }}
+	 hyperkube_hub: {{ .HiperkubeHub }}
```
Un-indent this.

luisyonaldo replied:
I think is github conmpare issue, can you check locally if you see the bad indentation too? locally I don't have it

ZackButcher replied:
Oh, tab-vs-spaces issue, I think! It shows up as a tab for me. YAML should always use spaces.

luisyonaldo replied:
thought I have spaces as default, indeed that's the reason

ZackButcher replied:
The fact that you're in a `.go` file probably threw your editor off, since golang uses tabs.

ZackButcher commented on install/kubernetes/helm/istio/charts/mixer/templates/create-custom-objects-job.yaml:
```diff
@@ -0,0 +1,35 @@
+    "helm.sh/hook-delete-policy": hook-succeeded
+  labels:
+    istio: mixer
+  name: mixer-create-custom-objects-job
```
[Pilot also has a whole swath of CRDs](https://github.com/istio/istio/blob/1c498e7712badd8f3e885fbe3b77e4d2d3621843/install/kubernetes/templates/istio-pilot.yaml.tmpl#L4-L45), we need to create them as well. Lets rename this "istio-create-custom-objects-job" and include them too.

ZackButcher replied:
nit: And maybe "custom-resources-job" instead of "custom-objects-job"?

luisyonaldo replied:
is inside `mixer` subchart (folder), pilot's CRD are inside `pilot` subchart. Also, didn't find any custom object for pilot that depend on CRD, that's why it doesn't need a post-hook job like in mixer chart.

ZackButcher replied:
Oh good point - yeah we don't create any Pilot resources at startup so there's no ordering requirements for Pilot CRDs.

istio-testing commented:
@luisyonaldo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | dca5c42b62b783b562ce6f400faf36f1c95d9097 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2461/istio-presubmit/2220/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

ZackButcher commented:
Presubmit is failing due to gen-deploy: you need to add `HyperkubeHub`/`HyperkubeTab` to the `installation` struct too.

istio-testing commented:
@luisyonaldo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | abb5ac1ff4b411174ce33f02843744b0cd051a82 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2461/istio-presubmit/2221/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@luisyonaldo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 81dbe6c85db1a714677b520ee3373529078d0231 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2461/istio-presubmit/2223/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

luisyonaldo commented:
@ZackButcher can you help me? Why is it failing now? Locally test pass

sdake commented:
@luisyonaldo I receive the following errors when deploying this chart:
```
sdake@falkor-07:~$ kubectl describe pod -n istio-system mixer-create-custom-resources-job-m66tv
Name:           mixer-create-custom-resources-job-m66tv
Namespace:      istio-system
Node:           falkor-07/10.23.230.11
Start Time:     Mon, 08 Jan 2018 16:57:24 -0700
Labels:         controller-uid=ac2b7a84-f4cf-11e7-95aa-003a7dd0301e
                istio=mixer
                job-name=mixer-create-custom-resources-job
Annotations:    kubernetes.io/created-by={"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"Job","namespace":"istio-system","name":"mixer-create-custom-resources-job","uid":"ac2b7a84-f4cf-11e7-95aa-...
Status:         Pending
IP:             192.168.104.118
Created By:     Job/mixer-create-custom-resources-job
Controlled By:  Job/mixer-create-custom-resources-job
Containers:
  hyperkube:
    Container ID:  
    Image:         map[HYPERKUBE_HUB:<nil>]:map[HYPERKUBE_TAG:<nil>]
```
I'd take a look at updateVersions.sh - perhaps that shell script needs a little bit of work.

Also, you may want to delete the file `deployIstioHelmCharts.sh` as with this change it is no longer valid.

ZackButcher commented:
The gendeploy test is failing:

```
I0108 22:21:57.425] --- FAIL: TestValuesFromInstallation (0.00s)
I0108 22:21:57.425]     --- FAIL: TestValuesFromInstallation/default (0.00s)
I0108 22:21:57.425]     	helm_test.go:36: valuesFromInstallation(&{true true true true false istio-system false 0 gcr.io/istio-testing 0.4.0 0.4.0 0.4.0 0.4.0 quay.io/coreos/hyperkube v1.7.6_coreos.0}) = "Istio:\n   deploy_base_config: true\n   initializer_enabled: false\n   mixer_enabled: true\n   pilot_enabled: true\n   ingress:\n      use_nodeport: false\n      nodeport_port: 0\n\nglobal:\n   auth_enabled: true\n   namespace: istio-system\n   ca_hub: gcr.io/istio-testing\n   ca_tag: 0.4.0\n   ca_demo: false\n   proxy_hub: gcr.io/istio-testing\n   proxy_tag: 0.4.0\n   proxy_debug: false\n   pilot_hub: gcr.io/istio-testing\n   pilot_tag: 0.4.0\n   hyperkube_hub: " wanted "Istio:\n   deploy_base_config: true\n   initializer_enabled: false\n   mixer_enabled: true\n   pilot_enabled: true\n   ingress:\n      use_nodeport: false\n      nodeport_port: 0\n\nglobal:\n   auth_enabled: true\n   namespace: istio-system\n   ca_hub: gcr.io/istio-testing\n   ca_tag: 0.4.0\n   ca_demo: false\n   proxy_hub: gcr.io/istio-testing\n   proxy_tag: 0.4.0\n   proxy_debug: false\n   pilot_hub: gcr.io/istio-testing\n   pilot_tag: 0.4.0\n   hyperkube_hub: quay.io/coreos/hyperkube\n   hyperkube_tag: v1.7.6_coreos.0\n"
I0108 22:21:57.425] FAIL
I0108 22:21:57.425] FAIL	istio.io/istio/pilot/cmd/istioctl/gendeployment	0.029s
```

Looks like there's a typo in the template so its not getting a value:
https://github.com/luisyonaldo/istio/blob/81dbe6c85db1a714677b520ee3373529078d0231/pilot/cmd/istioctl/gendeployment/helm.go#L43

linsun commented on pilot/cmd/istioctl/gendeployment/helm.go:
```diff
@@ -40,6 +40,8 @@ global:
    proxy_debug: {{ .Debug }}
    pilot_hub: {{ .Hub }}
    pilot_tag: {{ .PilotTag }}
+   hyperkube_hub: {{ .HiperkubeHub }}
```
This looks the typo to me:  HiperkubeHub -> . HyperkubeHub

linsun commented:
I can't get this chart deployed... I must be missing something really basic.   

$ helm install istio
Error: render error in "istio/templates/istio-ingress.yaml": template: istio/templates/istio-ingress.yaml:1:14: executing "istio/templates/istio-ingress.yaml" at <.Values.istio.deploy...>: can't evaluate field deploy_base_config in type interface {}

Is there a base config I should supply to helm when I install this chart?  I'm seeing same error with `deployIstioHelmCharts.sh`

linsun commented:
/test

sdake commented:
@linsun for me to deploy and receive the error described in `sdake@falkor-07:~$ kubectl describe pod -n istio-system mixer-create-custom-resources-job-m66tv`, I run against a freshly deployed Kubernetes cluster the following operations:
```
sdake@falkor-07~/go/src/istio.io/istio$ helm init (wait for tiller to enter RUNNING state)
sdake@falkor-07:~/go/src/istio.io/istio/install/kubernetes$ cd $ISTIO/istio/install/kubernetes/helm
sdake@falkor-07:~/go/src/istio.io/istio/install/kubernetes/helm$ kubectl create -f helm-service-account.yaml
helm install istio --name istio --set istio.ingress.use_nodeport=true
```

Note if you want to use a LoadBalancer service object type, you could instead run
```
helm install istio --name istio
```
in place of the last command.  using nodeport defaults to false.  Note I also ran updateVersions.sh with my own built version of Istio at HUB=docker.io/sdake and TAG=sdake as environment variables as per DEV-GUIDE.  I did commit to making a doc improvement on this work once it was functional.

The patch as is is currently broken however causing an image pull failure.  I'm sure @luisyonaldo can debug his work further to resolve this problem.  The `deployIstioHelmCharts.sh` file should be deleted and is incompatible with this change.

Cheers
-steve

luisyonaldo commented:
my bad, didn't add new values to `updateVersion.sh`, should work now, let me know

sdake commented:
testing now.

Cheers
-steve

luisyonaldo commented:
there is no `istio/charts/mixer/templates/istio-mixer.yaml` in the branch

sdake commented:
Yup user error.  I had a dirty repository.  I applied your patch with git am after cleaning up untracked files and receive the following result:
```
sdake@falkor-07:~/go/src/istio.io/istio/install/kubernetes/helm$ !helm
helm install istio --name istio --set istio.ingress.use_nodeport=true
Error: YAML parse error on istio/charts/pilot/templates/pilot.yaml: error converting YAML to JSON: yaml: line 17: mapping values are not allowed in this context
```

luisyonaldo commented:
not sure what's wrong, locally all works
```
$ helm install istio --name istio --set istio.ingress.use_nodeport=true
NAME:   istio
LAST DEPLOYED: Tue Jan  9 14:24:07 2018
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1beta1/ClusterRole
NAME                            AGE
istio-pilot-istio-system        13s
istio-ca-istio-system           13s
istio-initializer-istio-system  13s
istio-sidecar-istio-system      13s
istio-mixer-istio-system        13s

==> v1beta1/ClusterRoleBinding
NAME                                               AGE
istio-mixer-admin-role-binding-istio-system        13s
istio-ca-role-binding-istio-system                 13s
istio-sidecar-role-binding-istio-system            13s
istio-initializer-admin-role-binding-istio-system  13s
istio-ingress-admin-role-binding-istio-system      13s
istio-pilot-admin-role-binding-istio-system        13s

==> v1beta1/Deployment
NAME               DESIRED  CURRENT  UP-TO-DATE  AVAILABLE  AGE
istio-initializer  1        1        1           1          13s
istio-mixer        1        1        1           0          13s
istio-pilot        1        1        1           0          13s
istio-ingress      1        1        1           0          13s

==> v1/Namespace
NAME          STATUS  AGE
istio-system  Active  13s

==> v1/ConfigMap
NAME                              DATA  AGE
istio-inject                      1     13s
mixer-custom-resources-configmap  1     13s
istio-mixer                       1     13s
istio                             1     13s

==> v1beta1/CustomResourceDefinition
NAME                                 AGE
listcheckers.config.istio.io         13s
quotas.config.istio.io               13s
prometheuses.config.istio.io         13s
noops.config.istio.io                13s
rules.config.istio.io                13s
attributemanifests.config.istio.io   13s
deniers.config.istio.io              13s
memquotas.config.istio.io            13s
statsds.config.istio.io              13s
checknothings.config.istio.io        13s
kubernetesenvs.config.istio.io       13s
svcctrls.config.istio.io             13s
logentries.config.istio.io           13s
metrics.config.istio.io              13s
reportnothings.config.istio.io       13s
kuberneteses.config.istio.io         13s
stdios.config.istio.io               13s
listentries.config.istio.io          13s
stackdrivers.config.istio.io         13s
destinationpolicies.config.istio.io  13s
routerules.config.istio.io           13s
egressrules.config.istio.io          13s

==> v1/Pod(related)
NAME                            READY  STATUS             RESTARTS  AGE
istio-mixer-77c45fccc4-lsw84    0/3    ContainerCreating  0         13s
istio-pilot-5f64597fdc-7849z    0/2    ContainerCreating  0         13s
istio-ingress-6f8d69b45c-fjfkf  0/1    ContainerCreating  0         13s

==> v1/ServiceAccount
NAME                               SECRETS  AGE
istio-initializer-service-account  1        13s
istio-mixer-service-account        1        13s
istio-pilot-service-account        1        13s
istio-ingress-service-account      1        13s

==> v1/Service
NAME           TYPE       CLUSTER-IP      EXTERNAL-IP  PORT(S)                                                           AGE
istio-mixer    ClusterIP  10.97.8.166     <none>       9091/TCP,15004/TCP,9093/TCP,9094/TCP,9102/TCP,9125/UDP,42422/TCP  13s
istio-pilot    ClusterIP  10.99.218.105   <none>       15003/TCP,443/TCP                                                 13s
istio-ingress  NodePort   10.111.227.103  <none>       80:32000/TCP,443:32734/TCP                                        13s

==> v1alpha1/InitializerConfiguration
NAME           AGE
istio-sidecar  13s
```

sdake commented:
```
sdake@falkor-07:~/go/src/istio.io/istio/install/kubernetes/helm$ kubectl log -n istio-system istio-initializer-6659799985-87w75
W0109 06:29:43.419882   20390 cmd.go:392] log is DEPRECATED and will be removed in a future version. Use logs instead.
W0109 13:29:38.448205       1 client_config.go:529] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-01-09T13:29:38.449924Z	info	version sdake@falkor-07-4be2f37712a330948026f19778cfbeae93d88e1c-dirty-4be2f37712a330948026f19778cfbeae93d88e1c-dirty
Error: failed to read initializer configuration configmaps "istio-inject" is forbidden: User "system:serviceaccount:istio-system:istio-initializer-service-account" cannot get configmaps in the namespace "istio-system"
Usage:
  sidecar-initializer [flags]

Flags:
```

An initial look at helm-service-account.yaml looks like it is incorrectly specifying default rather than istio-system.  I'll report back.

What extended RBAC rules are you running?  We are almost there ;-)

Cheers
-steve

luisyonaldo commented:
I only added `create` permissions to `config.istio.io` api group https://github.com/istio/istio/pull/2461/files#diff-979305bddbcf79f15871f2c23206bca7 since now custom resources are created by post-install job which use `istio-mixer-service-account`

sbezverk commented:
@luisyonaldo Why do you use post-install hook? Have you tried with pre-install?

luisyonaldo commented:
as explained before, it doesn't matter if you define CRD in `pre-install` hook, helm validation step will fail if both CRD and CR are defined in the same install.

There is already an issue in helm, let me reference it.

luisyonaldo commented:
@sbezverk check this comment https://github.com/kubernetes/helm/issues/2925#issuecomment-328904829

sdake commented:
@luisyonaldo when I run with wide-open RBAC, i.e.:
```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: cluster-admin
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- kind: Group
  name: system:masters
- kind: Group
  name: system:authenticated
- kind: Group
  name: system:unauthenticated
```

I get the following results:
```
watch kubectl get pods --all-namespaces

Every 2.0s: kubectl get pods --all-namespaces                                                                                                                                                                                          Tue Jan  9 07:02:44 2018

NAMESPACE      NAME                                        READY     STATUS    RESTARTS   AGE
istio-system   istio-ingress-c46979744-wpgmt               1/1       Running   0          2m
istio-system   istio-initializer-6659799985-vqrfz          1/1       Running   0          2m
istio-system   istio-pilot-7796bf46c5-v7nnr                2/2       Running   0          2m
kube-system    calico-etcd-qvgvk                           1/1       Running   0          8m
kube-system    calico-node-ffgw2                           2/2       Running   1          8m
kube-system    calico-policy-controller-59fc4f7888-bjzg9   1/1       Running   0          8m
kube-system    etcd-falkor-07                              1/1       Running   0          7m
kube-system    kube-apiserver-falkor-07                    1/1       Running   0          7m
kube-system    kube-controller-manager-falkor-07           1/1       Running   0          7m
kube-system    kube-dns-545bc4bfd4-nbjfn                   3/3       Running   0          8m
kube-system    kube-proxy-wgj2k                            1/1       Running   0          8m
kube-system    kube-scheduler-falkor-07                    1/1       Running   0          7m
kube-system    tiller-deploy-5b9d65c7f-4dwx4               1/1       Running   0          3m
```

There doesn't appear to be a mixer pod.

Can you update this file `helm-service-account.yaml` to include the correct RBAC permissions?

Also, at a loss as to why mixer isn't deploying.

During deploy - my output looks like
```
sdake@falkor-07:~/go/src/istio.io/istio/install/kubernetes/helm$ !2361
helm install istio --name istio --set istio.ingress.use_nodeport=true
NAME:   istio
LAST DEPLOYED: Tue Jan  9 07:00:03 2018
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1/ConfigMap
NAME          DATA  AGE
istio-inject  1     1s
istio         1     1s

==> v1/ServiceAccount
NAME                               SECRETS  AGE
istio-initializer-service-account  1        1s
istio-pilot-service-account        1        1s
istio-ingress-service-account      1        1s

==> v1/Service
NAME           TYPE       CLUSTER-IP      EXTERNAL-IP  PORT(S)                     AGE
istio-pilot    ClusterIP  10.100.104.126  <none>       15003/TCP,443/TCP           1s
istio-ingress  NodePort   10.99.46.116    <none>       80:32000/TCP,443:31428/TCP  1s

==> v1beta1/Deployment
NAME               DESIRED  CURRENT  UP-TO-DATE  AVAILABLE  AGE
istio-initializer  1        1        1           0          1s
istio-pilot        1        1        1           0          1s
istio-ingress      1        1        1           0          1s

==> v1alpha1/InitializerConfiguration
NAME           AGE
istio-sidecar  1s

==> v1/Pod(related)
NAME                           READY  STATUS             RESTARTS  AGE
istio-pilot-7796bf46c5-v7nnr   0/2    ContainerCreating  0         1s
istio-ingress-c46979744-wpgmt  0/1    ContainerCreating  0         1s
```

Would you mind sharing your version of helm under test?

Cheers,
-steve

luisyonaldo commented:
sorry, I'm not using `helm-service-account.yaml`, I'm just using a SA with `cluster-admin` ClusterRole.

Example:
```
$ kubectl -n kube-system create sa tiller
$ kubectl create clusterrolebinding tiller --clusterrole cluster-admin --serviceaccount=kube-system:tiller
$ helm init --service-account tiller
```
let me try with `helm-service-account.yaml`.

Also, if you describe containers (`istio-mixer` and `istio-pilot`), you'll see they are failing when trying to pull image, this also happen if I tried normal steps on `master` right now, the `tags` seems to be wrong (I thought it was some access that I didn't have to debug with does images). I tested using release `v0.4.0` [values](https://github.com/istio/istio/blob/0.4.0/install/kubernetes/helm/istio/values.yaml)

sdake commented:
@luisyonaldo in the case of helm-service-account.yaml, mixer isn't spawned as a pod at all.  The Helm chart never renders it in the output (as shown in a previous comment).  I'll try your approach for SAs, maybe its better.

Cheers
-steve

sdake commented:
@luisyonaldo do you have a reference for installing helm via the above methodology?  (Looking for documentation to help hunt down what appears to be an RBAC issue in my environment).

I tried the approach you mentioned, and receive the following error:

```
sdake@falkor-07:~/go/src/istio.io/istio/install/kubernetes/helm$ kubectl log istio-initializer-6659799985-b8492 -n istio-system
W0109 08:37:47.541813   31292 cmd.go:392] log is DEPRECATED and will be removed in a future version. Use logs instead.
W0109 15:37:27.451866       1 client_config.go:529] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-01-09T15:37:27.453199Z	info	version sdake@falkor-07-4be2f37712a330948026f19778cfbeae93d88e1c-dirty-4be2f37712a330948026f19778cfbeae93d88e1c-dirty
Error: failed to read initializer configuration configmaps "istio-inject" is forbidden: User "system:serviceaccount:istio-system:istio-initializer-service-account" cannot get configmaps in the namespace "istio-system"
Usage:
  sidecar-initializer [flags]
```

What type of Kubernetes environment are you deploying into?  Are you deploying any other special RBAC rules?

I have no doubt you have a working system - I'd like to replicate it so others can also replicate it (via documentation).  FWIW, I am running with no RBAC rules and deploying via kubeadm.  The kubeadm config file I use is in this review.

Finally, mixer still fails to render and launch:

```
sdake@falkor-07:~/go/src/istio.io/istio/install/kubernetes/helm$ !1921
helm install istio --name istio --set istio.ingress.use_nodeport=true
NAME:   istio
LAST DEPLOYED: Tue Jan  9 08:35:47 2018
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1/ConfigMap
NAME          DATA  AGE
istio-inject  1     1s
istio         1     1s

==> v1/ServiceAccount
NAME                               SECRETS  AGE
istio-initializer-service-account  1        1s
istio-pilot-service-account        1        1s
istio-ingress-service-account      1        1s

==> v1/Service
NAME           TYPE       CLUSTER-IP     EXTERNAL-IP  PORT(S)                     AGE
istio-pilot    ClusterIP  10.99.151.186  <none>       15003/TCP,443/TCP           1s
istio-ingress  NodePort   10.102.61.197  <none>       80:32000/TCP,443:32348/TCP  1s

==> v1beta1/Deployment
NAME               DESIRED  CURRENT  UP-TO-DATE  AVAILABLE  AGE
istio-initializer  1        1        1           0          1s
istio-pilot        1        1        1           0          1s
istio-ingress      1        1        1           0          1s

==> v1alpha1/InitializerConfiguration
NAME           AGE
istio-sidecar  1s

==> v1/Pod(related)
NAME                           READY  STATUS             RESTARTS  AGE
istio-pilot-7796bf46c5-xt8gj   0/2    ContainerCreating  0         1s
istio-ingress-c46979744-tx59q  0/1    ContainerCreating  0         1s
```

sdake commented:
Docs on RBAC with Helm (these didn't exist when I last used helm ;)

https://github.com/kubernetes/helm/blob/master/docs/rbac.md

luisyonaldo commented:
helm [documentation](https://docs.helm.sh/using_helm/#tiller-and-role-based-access-control) to use on a cluster with RBAC. If you already have installed tiller then run `helm init --upgrade --service-account=tiller` or run `helm reset` before installing it again.

I'm testing locally with minikube (both with RBAC enabled and disabled) and also on self-hosted 8 node cluster deployed with kubeadmin and RBAC enabled. On both cases there are no issues.

Also, I think issue is with images tags, I have same issue. If I run `updateVersion.sh` the `tags` it generated for `` fail when you try to pull.
```
NAME                                    READY     STATUS             RESTARTS   AGE
po/istio-ingress-6f8d69b45c-5lxz5       0/1       ImagePullBackOff   0          4m
po/istio-initializer-5cd7fb95f5-mcxz4   1/1       Running            0          4m
po/istio-mixer-77c45fccc4-hjttc         1/3       ImagePullBackOff   5          4m
po/istio-pilot-5f64597fdc-5vm7r         1/2       ImagePullBackOff   0          4m
```
for me only `v0.4.0` release values work.

linsun commented:
@luisyonaldo do you have CLA signed?

linsun commented:
/lgtm

sdake commented:
@linsun i'd encourage one more change on this PR.  I was able to deploy the work with 1 command - but had to manually modify a file that updateVersions.sh should be modifying.  @luisyonaldo has done a great job with this work and we spoke.  He indicated he can wrap up the updateVersions.sh problem quickly.

Cheers
-steve

linsun commented:
@sdake glad it works for you!   That is fine if it is in fast enough.  Else, a separate pr is fine too.  Smaller PR is easier to review and recommended in general.

sdake commented:
@linsun sounds good.  Lets get it merged today so I can get the docs up on istio.io.

Cheers
-steve

ZackButcher commented:
I can approve and we can merge as soon as the last updateVersions.sh bit is in.

sdake commented:
This patch is GTG.  Please feel free to merge  👍 Rockin job @luisyonaldo !

ZackButcher commented:
/approve
/lgtm

linsun commented:
@luisyonaldo I was able to run through everything too, nice work!
